crossdock/client/ctxpropagation/behavior.go

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ctxpropagation
+
+import (
+	js "encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/yarpc/yarpc-go"
+	"github.com/yarpc/yarpc-go/crossdock-go"
+	"github.com/yarpc/yarpc-go/crossdock/client/params"
+	server "github.com/yarpc/yarpc-go/crossdock/server/yarpc"
+	"github.com/yarpc/yarpc-go/encoding/json"
+	"github.com/yarpc/yarpc-go/internal/baggage"
+	"github.com/yarpc/yarpc-go/transport"
+	ht "github.com/yarpc/yarpc-go/transport/http"
+	tch "github.com/yarpc/yarpc-go/transport/tchannel"
+
+	"github.com/uber/tchannel-go"
+	"golang.org/x/net/context"
+)
+
+type handler interface {
+	Inject(json.Client, server.TransportConfig) // this is hacky but whatever
+	Handle(*json.ReqMeta, interface{}) (interface{}, *json.ResMeta, error)
+}
+
+func assertBaggageMatches(t crossdock.T, ctx context.Context, want transport.Headers) bool {
+	assert := crossdock.Assert(t)
+	got := baggage.FromContext(ctx)
+	if len(want) == 0 { // nils
```

Comment:
That we're doing a `len` check here to gracefully handle `nil` maps. `reflect.DeepEqual(nil, transport.Headers{}) == false`, but `len(nil) == len(transport.Headers{})`.

I'll update the comment.